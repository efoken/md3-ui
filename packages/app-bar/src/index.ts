export * from "./app-bar"
export * from "./toolbar"
