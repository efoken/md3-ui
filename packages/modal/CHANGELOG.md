# Change Log

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.2.0](https://github.com/efoken/md3-ui/compare/v0.1.0...v0.2.0) (2022-06-29)

### Features

- add MediaQuery styles via classes instead of data attributes
  ([0df0626](https://github.com/efoken/md3-ui/commit/0df0626be02ccd28c589de1ea0a71630dc0ddadd))

# Change Log

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.1.0](https://github.com/efoken/md3-ui/compare/v0.0.5...v0.1.0) (2022-06-24)

**Note:** Version bump only for package @md3-ui/modal
