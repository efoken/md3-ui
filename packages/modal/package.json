{
  "name": "@md3-ui/modal",
  "version": "0.0.3",
  "main": "dist/md3-ui-modal.cjs.js",
  "license": "MIT",
  "homepage": "https://eikefoken.com/md3-ui/",
  "repository": {
    "type": "git",
    "url": "https://github.com/efoken/md3-ui",
    "directory": "packages/modal"
  },
  "keywords": [
    "react",
    "react-component",
    "material design",
    "material you"
  ],
  "files": [
    "dist",
    "src"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@md3-ui/hooks": "^0.0.3",
    "@md3-ui/portal": "^0.0.3",
    "@md3-ui/system": "^0.0.3",
    "@md3-ui/utils": "^0.0.3"
  },
  "devDependencies": {
    "@types/react": "^17.0.44",
    "@types/react-dom": "^17.0.15",
    "@types/react-native": "^0.67.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-native": "^0.68.0"
  },
  "peerDependencies": {
    "@types/react": "^17.0.44 || ^18.0.0",
    "@types/react-dom": "^17.0.15 || ^18.0.0",
    "@types/react-native": "^0.67.3",
    "react": "^17.0.2 || ^18.0.0",
    "react-dom": "^17.0.2 || ^18.0.0",
    "react-native": ">=0.67.4"
  }
}
