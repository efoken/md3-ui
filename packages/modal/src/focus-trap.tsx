import { useForkRef } from "@md3-ui/hooks"
import { getOwnerDocument, runIfFn, __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import { Platform } from "react-native"

// Inspired by https://github.com/focus-trap/tabbable
const candidatesSelector = [
  "input",
  "select",
  "textarea",
  "a[href]",
  "button",
  "[tabindex]",
  "audio[controls]",
  "video[controls]",
  '[contenteditable]:not([contenteditable="false"])',
].join(",")

export interface FocusTrapProps {
  /**
   * A single child content element.
   */
  children: React.ReactElement
  /**
   * If `true`, the focus trap will not automatically shift focus to itself when
   * it opens, and replace it to the last focused element when it closes.
   * This also works correctly with any focus trap children that have the
   * `disableAutoFocus` prop.
   *
   * Generally this should never be set to `true` as it makes the focus trap
   * less accessible to assistive technologies, like screen readers.
   * @default false
   */
  disableAutoFocus?: boolean
  /**
   * If `true`, the focus trap will not prevent focus from leaving the focus
   * trap while open.
   *
   * Generally this should never be set to `true` as it makes the focus trap
   * less accessible to assistive technologies, like screen readers.
   * @default false
   */
  disableEnforceFocus?: boolean
  /**
   * If `true`, the focus trap will not restore focus to previously focused
   * element once focus trap is hidden or unmounted.
   * @default false
   */
  disableRestoreFocus?: boolean
  /**
   * This prop extends the `open` prop. It allows to toggle the open state
   * without having to wait for a rerender when changing the `open` prop. This
   * prop should be memoized. It can be used to support multiple focus trap
   * mounted at the same time.
   * @default true
   */
  enabled?: boolean | (() => boolean)
  /**
   * Returns an array of ordered tabbable nodes (i.e. in tab order) within the
   * root. For instance, you can provide the "tabbable" npm dependency.
   */
  getTabbable?: (root: HTMLElement) => HTMLElement[]
  /**
   * If `true`, focus is locked.
   */
  open: boolean
}

function getTabIndex(node: HTMLElement) {
  const tabindexAttr = Number.parseInt(node.getAttribute("tabindex") ?? "0", 10)

  if (!Number.isNaN(tabindexAttr)) {
    return tabindexAttr
  }

  // Browsers do not return `tabIndex` correctly for contentEditable nodes;
  // https://bugs.chromium.org/p/chromium/issues/detail?id=661108&q=contenteditable%20tabindex&can=2
  // so if they don't have a tabindex attribute specifically set, assume it's 0.
  // In Chrome, <details />, <audio controls /> and <video controls /> elements
  // get a default `tabIndex` of -1 when the 'tabindex' attribute isn't
  // specified in the DOM, yet they are still part of the regular tab order;
  // in FF, they get a default `tabIndex` of 0; since Chrome still puts those
  // elements in the regular tab order, consider their tab index to be 0.
  if (
    node.contentEditable === "true" ||
    ((node.nodeName === "AUDIO" ||
      node.nodeName === "VIDEO" ||
      node.nodeName === "DETAILS") &&
      node.getAttribute("tabindex") == null)
  ) {
    return 0
  }

  return node.tabIndex
}

function isNonTabbableRadio(node: HTMLElement): node is HTMLInputElement {
  if (node.tagName !== "INPUT" || (node as HTMLInputElement).type !== "radio") {
    return false
  }

  if (!(node as HTMLInputElement).name) {
    return false
  }

  const getRadio = (selector: string) =>
    node.ownerDocument.querySelector(`input[type="radio"]${selector}`)

  let roving = getRadio(`[name="${(node as HTMLInputElement).name}"]:checked`)

  if (!roving) {
    roving = getRadio(`[name="${(node as HTMLInputElement).name}"]`)
  }

  return roving !== node
}

function isNodeMatchingSelectorFocusable(node: HTMLElement) {
  if (
    (node as HTMLInputElement).disabled ||
    (node.tagName === "INPUT" &&
      (node as HTMLInputElement).type === "hidden") ||
    isNonTabbableRadio(node)
  ) {
    return false
  }
  return true
}

function defaultGetTabbable(root: HTMLElement) {
  const regularTabNodes: HTMLElement[] = []
  const orderedTabNodes: {
    documentOrder: number
    node: HTMLElement
    tabIndex: number
  }[] = []

  for (const [i, node] of root
    .querySelectorAll<HTMLElement>(candidatesSelector)
    .entries()) {
    const nodeTabIndex = getTabIndex(node)

    if (nodeTabIndex === -1 || !isNodeMatchingSelectorFocusable(node)) {
      continue
    }

    if (nodeTabIndex === 0) {
      regularTabNodes.push(node)
    } else {
      orderedTabNodes.push({
        documentOrder: i,
        node,
        tabIndex: nodeTabIndex,
      })
    }
  }

  return [
    ...orderedTabNodes
      .sort((a, b) =>
        a.tabIndex === b.tabIndex
          ? a.documentOrder - b.documentOrder
          : a.tabIndex - b.tabIndex,
      )
      .map((a) => a.node),
    ...regularTabNodes,
  ]
}

/**
 * Utility component that locks focus inside the component.
 */
export const FocusTrap: React.FC<FocusTrapProps> = ({
  children,
  disableAutoFocus = false,
  disableEnforceFocus = false,
  disableRestoreFocus = false,
  enabled = true,
  getTabbable = defaultGetTabbable,
  open,
}) => {
  const ignoreNextEnforceFocus = React.useRef<boolean>()
  const sentinelStartRef = React.useRef<HTMLDivElement>(null)
  const sentinelEndRef = React.useRef<HTMLDivElement>(null)
  const restoreTarget = React.useRef<HTMLElement | null>(null)
  const focusEventTarget = React.useRef<HTMLElement | null>(null)
  // This variable is useful when `disableAutoFocus` is `true`.
  // It waits for the active element to move into the component to activate.
  const activated = React.useRef(false)

  const rootRef = React.useRef<HTMLElement>(null)
  const handleRef = useForkRef((children as any).ref, rootRef)

  const lastKeydown = React.useRef<KeyboardEvent>()

  React.useEffect(() => {
    // We might render an empty child.
    if (!open || rootRef.current == null) {
      return
    }
    activated.current = !disableAutoFocus
  }, [disableAutoFocus, open])

  React.useEffect(() => {
    // We might render an empty child.
    if (!open || rootRef.current == null) {
      return () => {}
    }

    const doc = getOwnerDocument(rootRef.current)

    if (!rootRef.current.contains(doc.activeElement)) {
      if (!rootRef.current.hasAttribute("tabIndex")) {
        if (__DEV__) {
          console.error(
            [
              "MD3-UI: The modal content node does not accept focus.",
              "For the benefit of assistive technologies, " +
                'the tabIndex of the node is being set to "-1".',
            ].join("\n"),
          )
        }
        rootRef.current.setAttribute("tabIndex", "-1")
      }

      if (activated.current) {
        rootRef.current.focus()
      }
    }

    return () => {
      if (!disableRestoreFocus) {
        if (restoreTarget.current != null) {
          ignoreNextEnforceFocus.current = true
          restoreTarget.current.focus()
        }
        restoreTarget.current = null
      }
    }
    // Missing `disableRestoreFocus` which is fine.
    // We don't support changing that prop on an open FocusTrap
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [open])

  React.useEffect(() => {
    // We might render an empty child.
    if (!open || rootRef.current == null) {
      return () => {}
    }

    const doc = getOwnerDocument(rootRef.current)

    const contain = (nativeEvent?: FocusEvent) => {
      const { current: rootElement } = rootRef

      // Cleanup functions are executed lazily in React 17.
      // So `contain` can be called between the component being unmounted and
      // its cleanup function being run.
      if (rootElement == null) {
        return
      }

      if (
        !doc.hasFocus() ||
        disableEnforceFocus ||
        !runIfFn(enabled) ||
        ignoreNextEnforceFocus.current
      ) {
        ignoreNextEnforceFocus.current = false
        return
      }

      if (!rootElement.contains(doc.activeElement)) {
        // If the focus event is not coming from inside the children's React
        // tree, reset the refs.
        if (
          (nativeEvent && focusEventTarget.current !== nativeEvent.target) ||
          doc.activeElement !== focusEventTarget.current
        ) {
          focusEventTarget.current = null
        } else if (focusEventTarget.current !== null) {
          return
        }

        if (!activated.current) {
          return
        }

        let tabbable: HTMLElement[] = []
        if (
          doc.activeElement === sentinelStartRef.current ||
          doc.activeElement === sentinelEndRef.current
        ) {
          tabbable = getTabbable(rootElement)
        }

        if (tabbable.length > 0) {
          const isShiftTab = Boolean(
            lastKeydown.current?.shiftKey && lastKeydown.current?.key === "Tab",
          )

          const focusNext = tabbable[0]
          const focusPrevious = tabbable[tabbable.length - 1]

          if (isShiftTab) {
            focusPrevious.focus()
          } else {
            focusNext.focus()
          }
        } else {
          rootElement.focus()
        }
      }
    }

    const loopFocus = (nativeEvent: KeyboardEvent) => {
      lastKeydown.current = nativeEvent

      if (
        disableEnforceFocus ||
        !runIfFn(enabled) ||
        nativeEvent.key !== "Tab"
      ) {
        return
      }

      // Make sure the next tab starts from the right place.
      // `doc.activeElement` referes to the origin.
      if (doc.activeElement === rootRef.current && nativeEvent.shiftKey) {
        // We need to ignore the next `contain`, as it will try to move the
        // focus back to the `rootRef` element.
        ignoreNextEnforceFocus.current = true
        sentinelEndRef.current?.focus()
      }
    }

    doc.addEventListener("focusin", contain)
    doc.addEventListener("keydown", loopFocus, true)

    // With Edge, Safari and Firefox, no focus related events are fired when the
    // focused area stops being a focused area
    // (e.g. https://bugzilla.mozilla.org/show_bug.cgi?id=559561).
    // Instead, we can look if the active element was restored on the <body>
    // element.
    //
    // The WHATWG spec defines how the browser should behave but does not
    // explicitly mention any events:
    // https://html.spec.whatwg.org/multipage/interaction.html#focus-fixup-rule.
    const interval = setInterval(() => {
      if (doc.activeElement?.tagName === "BODY") {
        contain()
      }
    }, 50)

    return () => {
      clearInterval(interval)

      doc.removeEventListener("focusin", contain)
      doc.removeEventListener("keydown", loopFocus, true)
    }
  }, [
    disableAutoFocus,
    disableEnforceFocus,
    disableRestoreFocus,
    enabled,
    getTabbable,
    open,
  ])

  const handleFocus = (event: React.FocusEvent<HTMLDivElement>) => {
    if (restoreTarget.current == null) {
      restoreTarget.current = event.relatedTarget as HTMLElement
    }
    activated.current = true
    focusEventTarget.current = event.target as HTMLElement

    const childrenPropsHandler = children.props.onFocus
    if (childrenPropsHandler) {
      childrenPropsHandler(event)
    }
  }

  const handleFocusSentinel = (event: React.FocusEvent<HTMLDivElement>) => {
    if (restoreTarget.current == null) {
      restoreTarget.current = event.relatedTarget as HTMLElement
    }
    activated.current = true
  }

  return Platform.OS === "web" ? (
    <>
      <div
        ref={sentinelStartRef}
        data-testid="sentinel-start"
        // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        tabIndex={0}
        onFocus={handleFocusSentinel}
      />
      {React.cloneElement(children, { ref: handleRef, onFocus: handleFocus })}
      <div
        ref={sentinelEndRef}
        data-testid="sentinel-end"
        // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        tabIndex={0}
        onFocus={handleFocusSentinel}
      />
    </>
  ) : (
    children
  )
}

if (__DEV__) {
  FocusTrap.displayName = "FocusTrap"
}
