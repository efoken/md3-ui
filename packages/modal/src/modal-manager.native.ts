export interface ManagedModalProps {
  disableScrollLock?: boolean
}

export const ModalManager = {
  add: () => 0,
  mount: () => {},
  remove: () => 0,
  isTopModal: () => true,
}
