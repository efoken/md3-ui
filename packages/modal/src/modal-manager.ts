import { getOwnerDocument, getOwnerWindow } from "@md3-ui/utils"

interface ManagedModalProps {
  disableScrollLock?: boolean
}

function getScrollbarSize(doc: Document) {
  // https://developer.mozilla.org/en-US/docs/Web/API/Window/innerWidth#usage_notes
  return Math.abs(window.innerWidth - doc.documentElement.clientWidth)
}

// Is a vertical scrollbar displayed?
function isOverflowing(container: Element) {
  const doc = getOwnerDocument(container)

  if (doc.body === container) {
    return (
      getOwnerWindow(container).innerWidth > doc.documentElement.clientWidth
    )
  }

  return container.scrollHeight > container.clientHeight
}

function ariaHidden(element: Element, show: boolean) {
  if (show) {
    element.setAttribute("aria-hidden", "true")
  } else {
    element.removeAttribute("aria-hidden")
  }
}

function getPaddingRight(element: Element) {
  return Number.parseInt(
    getOwnerWindow(element).getComputedStyle(element).paddingRight ?? "0",
    10,
  )
}

function ariaHiddenSiblings(
  container: Element,
  mountElement: Element,
  currentElement: Element,
  elementsToExclude: readonly Element[] = [],
  show = false,
): void {
  const blacklist = new Set([
    mountElement,
    currentElement,
    ...elementsToExclude,
  ])
  const blacklistTagNames = new Set(["TEMPLATE", "SCRIPT", "STYLE"])

  for (const element of Array.from(container.children)) {
    if (!blacklist.has(element) && !blacklistTagNames.has(element.tagName)) {
      ariaHidden(element, show)
    }
  }
}

function handleContainer(containerInfo: Container, props: ManagedModalProps) {
  const restoreStyle: {
    el: HTMLElement | SVGElement
    property: string
    value: string
  }[] = []
  const { container } = containerInfo

  if (!props.disableScrollLock) {
    if (isOverflowing(container)) {
      // Compute the size before applying `overflow: hidden` to avoid any
      // scroll jumps.
      const scrollbarSize = getScrollbarSize(getOwnerDocument(container))

      restoreStyle.push({
        el: container,
        property: "padding-right",
        value: container.style.paddingRight,
      })
      // Use computed style here, to get the real padding to add our scrollbar
      // width to.
      container.style.paddingRight = `${
        getPaddingRight(container) + scrollbarSize
      }px`

      // `.md3-fixed` is a global helper.
      const fixedElements = getOwnerDocument(container).querySelectorAll<
        HTMLElement | SVGElement
      >(".md3-fixed")

      for (const element of fixedElements) {
        restoreStyle.push({
          el: element,
          property: "padding-right",
          value: element.style.paddingRight,
        })
        element.style.paddingRight = `${
          getPaddingRight(element) + scrollbarSize
        }px`
      }
    }

    // Improve Gatsby support
    // https://css-tricks.com/snippets/css/force-vertical-scrollbar/
    const parent = container.parentElement
    const containerWindow = getOwnerWindow(container)
    const scrollContainer =
      parent?.nodeName === "HTML" &&
      containerWindow.getComputedStyle(parent).overflowY === "scroll"
        ? parent
        : container

    // Block the scroll even if no scrollbar is visible to account for mobile
    // keyboard screensize shrink.
    restoreStyle.push(
      {
        el: scrollContainer,
        property: "overflow",
        value: scrollContainer.style.overflow,
      },
      {
        el: scrollContainer,
        property: "overflow-x",
        value: scrollContainer.style.overflowX,
      },
      {
        el: scrollContainer,
        property: "overflow-y",
        value: scrollContainer.style.overflowY,
      },
    )

    scrollContainer.style.overflow = "hidden"
  }

  const restore = () => {
    for (const { value, el, property } of restoreStyle) {
      if (value) {
        el.style.setProperty(property, value)
      } else {
        el.style.removeProperty(property)
      }
    }
  }

  return restore
}

function getHiddenSiblings(container: Element) {
  return Array.from(container.children).filter(
    (element) => element.getAttribute("aria-hidden") === "true",
  )
}

interface Modal {
  modalRef: Element
  mountNode: Element
}

interface Container {
  container: HTMLElement
  hiddenSiblings: Element[]
  modals: Modal[]
  restore: null | (() => void)
}

/**
 * Proper state management for containers and the modals in those containers.
 * Simplified, but inspired by react-overlay's ModalManager class. Used by the
 * Modal to ensure proper styling of containers.
 */
export const ModalManager = new (class ModalManager {
  containers: Container[]

  modals: Modal[]

  constructor() {
    this.modals = []
    this.containers = []
  }

  add(modal: Modal, container: HTMLElement) {
    let modalIndex = this.modals.indexOf(modal)
    if (modalIndex !== -1) {
      return modalIndex
    }

    modalIndex = this.modals.length
    this.modals.push(modal)

    // If the modal we are adding is already in the DOM.
    if (modal.modalRef) {
      ariaHidden(modal.modalRef, false)
    }

    const hiddenSiblings = getHiddenSiblings(container)
    ariaHiddenSiblings(
      container,
      modal.mountNode,
      modal.modalRef,
      hiddenSiblings,
      true,
    )

    const containerIndex = this.containers.findIndex(
      (item) => item.container === container,
    )
    if (containerIndex !== -1) {
      this.containers[containerIndex].modals.push(modal)
      return modalIndex
    }

    this.containers.push({
      modals: [modal],
      container,
      restore: null,
      hiddenSiblings,
    })

    return modalIndex
  }

  mount(modal: Modal, props: ManagedModalProps) {
    const containerInfo = this.containers.find((item) =>
      item.modals.includes(modal),
    )

    if (containerInfo != null && containerInfo.restore == null) {
      containerInfo.restore = handleContainer(containerInfo, props)
    }
  }

  remove(modal: Modal) {
    const modalIndex = this.modals.indexOf(modal)

    if (modalIndex === -1) {
      return modalIndex
    }

    const containerIndex = this.containers.findIndex((item) =>
      item.modals.includes(modal),
    )
    const containerInfo = this.containers[containerIndex]

    containerInfo.modals.splice(containerInfo.modals.indexOf(modal), 1)
    this.modals.splice(modalIndex, 1)

    // If that was the last modal in a container, clean up the container.
    if (containerInfo.modals.length === 0) {
      // The modal might be closed before it had the chance to be mounted in the DOM.
      containerInfo.restore?.()

      if (modal.modalRef) {
        // In case the modal wasn't in the DOM yet.
        ariaHidden(modal.modalRef, true)
      }

      ariaHiddenSiblings(
        containerInfo.container,
        modal.mountNode,
        modal.modalRef,
        containerInfo.hiddenSiblings,
        false,
      )
      this.containers.splice(containerIndex, 1)
    } else {
      // Otherwise make sure the next top modal is visible to a screen reader.
      const nextTop = containerInfo.modals[containerInfo.modals.length - 1]
      // As soon as a modal is adding its modalRef is undefined. it can't set
      // aria-hidden because the dom element doesn't exist either
      // when modal was unmounted before modalRef gets null
      if (nextTop.modalRef) {
        ariaHidden(nextTop.modalRef, false)
      }
    }

    return modalIndex
  }

  isTopModal(modal: Modal) {
    return (
      this.modals.length > 0 && this.modals[this.modals.length - 1] === modal
    )
  }
})()
