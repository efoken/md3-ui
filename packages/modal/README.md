# @md3-ui/modal

## Installation

```sh
// with yarn
yarn add @md3-ui/modal

// with npm
npm i @md3-ui/modal
```

## Import Components

```jsx
import { Modal } from "@md3-ui/modal"
```
