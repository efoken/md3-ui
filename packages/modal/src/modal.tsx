import {
  useAnimate,
  useBackHandler,
  useEventCallback,
  useForkRef,
} from "@md3-ui/hooks"
import { Portal } from "@md3-ui/portal"
import {
  OverridableComponent,
  OverrideProps,
  OwnerStateProps,
  styled,
  SxProps,
  useThemeProps,
} from "@md3-ui/system"
import { getOwnerDocument, __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  Animated,
  Easing,
  Platform,
  TouchableWithoutFeedback,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"
import { FocusTrap, FocusTrapProps } from "./focus-trap"
import { ModalManager } from "./modal-manager"

export interface ModalTypeMap<
  P = {},
  C extends React.ElementType = typeof RNView,
> {
  props: P &
    Pick<
      FocusTrapProps,
      "disableAutoFocus" | "disableEnforceFocus" | "disableRestoreFocus"
    > & {
      /**
       * A single child content element.
       */
      children: React.ReactElement
      /**
       * The ref to the component that will have the portal children appended to
       * it. By default, it uses the body of the top-level document object.
       */
      containerRef?: React.RefObject<any>
      /**
       * If `true`, hitting escape will not fire the `onClose` callback.
       * @default false
       */
      disableEscapeKeyDown?: boolean
      /**
       * If `true`, the `children` will be under the DOM hierarchy of the parent
       * component.
       * @default false
       */
      disablePortal?: boolean
      /**
       * Disable the scroll lock behavior.
       * @default false
       */
      disableScrollLock?: boolean
      /**
       * If `true`, the scrim is not rendered.
       * @default false
       */
      hideScrim?: boolean
      /**
       * Always keep the children in the DOM. This prop can be useful in SEO
       * situation or when you want to maximize the responsiveness of the Modal.
       * @default false
       */
      keepMounted?: boolean
      /**
       * Callback fired when the component requests to be closed.
       */
      onClose?: () => void
      /**
       * If `true`, the component is shown.
       */
      open: boolean
      /**
       * Override or extend the styles applied to the component.
       */
      styles?: {
        root?: RNViewStyle
      }
      /**
       * The system prop that allows defining system overrides as well as
       * additional styles.
       */
      sx?: SxProps
    }
  defaultAs: C
}

export type ModalProps<
  C extends React.ElementType = ModalTypeMap["defaultAs"],
  P = {},
> = OverrideProps<ModalTypeMap<P, C>, C>

export type ModalStyleKey = keyof NonNullable<ModalProps["styles"]>

const ModalRoot = styled(RNView, {
  name: "Modal",
  slot: "Root",
})<OwnerStateProps<Pick<ModalProps, "open"> & { exited: boolean }>>(
  ({ theme, ownerState }) => ({
    bottom: 0,
    left: 0,
    position: Platform.OS === "web" ? ("fixed" as any) : "absolute",
    right: 0,
    top: 0,
    zIndex: theme.zIndex.modal,
    ...(!ownerState.open &&
      ownerState.exited &&
      Platform.select({
        web: {
          visibility: "hidden",
        },
        default: {
          display: "none",
        },
      })),
  }),
)

const ModalScrim = styled(Animated.View, {
  name: "Modal",
  slot: "Scrim",
  skipSx: true,
})(({ theme }) => ({
  backgroundColor: theme.utils.rgba("#322f37", 0.4),
  height: "100%",
  position: "absolute",
  width: "100%",
  zIndex: -1,
}))

export const Modal = React.forwardRef<RNView, ModalProps>((inProps, ref) => {
  const {
    children,
    containerRef,
    disableAutoFocus,
    disableEnforceFocus,
    disableEscapeKeyDown = false,
    disablePortal = false,
    disableRestoreFocus,
    disableScrollLock = false,
    hideScrim = false,
    keepMounted = false,
    onClose,
    onKeyDown,
    open,
    style,
    styles,
    ...props
  } = useThemeProps({
    name: "Modal",
    props: inProps,
  })

  const mountNodeRef = React.useRef<any>(null)
  const rootRef = React.useRef<RNView>(null)
  const handleRef = useForkRef(rootRef, ref)

  const modalRef = React.useRef({
    modalRef: rootRef.current as unknown as HTMLElement,
    mountNode: mountNodeRef.current,
  })

  const [exited, setExited] = React.useState(true)
  const [hidden, setHidden] = React.useState(false)

  const [opacity, { start: animate }] = useAnimate({
    duration: 225,
    easing: Easing.bezier(0.4, 0, 0.2, 1),
    fromValue: open ? 0 : 1,
    shouldReset: false,
    toValue: open ? 1 : 0,
    useNativeDriver: true,
  })

  const getDocument = () => getOwnerDocument(mountNodeRef.current)

  const getModal = () => {
    modalRef.current.modalRef = rootRef.current as unknown as HTMLElement
    modalRef.current.mountNode = mountNodeRef.current
    return modalRef.current
  }

  const isTopModal = () => ModalManager.isTopModal(getModal())

  const handleMounted = () => {
    ModalManager.mount(getModal(), { disableScrollLock })
    setHidden(false)

    // Fix a bug on Chrome where the scroll isn't initially 0.
    ;(rootRef.current as unknown as HTMLElement).scrollTop = 0
  }

  const handlePortalRef = (node: any) => {
    mountNodeRef.current = node

    if (node == null) {
      return
    }

    if (open && isTopModal()) {
      handleMounted()
    } else {
      setHidden(true)
    }
  }

  const handleOpen = useEventCallback(() => {
    const container =
      (containerRef?.current as HTMLElement | null) ?? getDocument().body

    animate((finished) => {
      if (finished) {
        setExited(false)
      }
    })
    ModalManager.add(getModal(), container)
    setHidden(false)

    // The element was already mounted.
    if (rootRef.current != null) {
      handleMounted()
    }
  })

  const handleClose = useEventCallback(() => {
    animate((finished) => {
      if (finished) {
        setExited(true)
      }
    })
    ModalManager.remove(getModal())
    setHidden(true)
  })

  React.useEffect(() => {
    if (open) {
      setExited(false)
      handleOpen()
    } else {
      handleClose()
    }
  }, [handleClose, handleOpen, open])

  // Close the modal on unmount
  React.useEffect(() => handleClose, [handleClose])

  useBackHandler(() => {
    onClose?.()
    return true
  })

  if (!keepMounted && !open && exited) {
    return null
  }

  const handleKeyDown = (event: React.KeyboardEvent) => {
    onKeyDown?.(event)

    // The handler doesn't take `event.defaultPrevented` into account:
    //
    // `event.preventDefault()` is meant to stop default behaviors like pressing
    // a checkbox to check it, hitting a button to submit a form, and hitting
    // left arrow to move the cursor in a text input etc.
    // Only special HTML elements have these default behaviors.
    if (event.key !== "Escape" || !isTopModal()) {
      return
    }

    if (!disableEscapeKeyDown) {
      // Swallow the event, in case someone is listening for the escape key on
      // the body.
      event.stopPropagation()

      onClose?.()
    }
  }

  const ownerState = {
    exited,
    open,
  }

  return (
    <Portal
      ref={handlePortalRef}
      containerRef={containerRef}
      disablePortal={disablePortal}
    >
      <ModalRoot
        ref={handleRef}
        accessibilityViewIsModal
        accessibilityRole="none"
        ownerState={ownerState}
        {...(hidden && {
          accessibilityElementsHidden: true,
          accessibilityHidden: true,
          importantForAccessibility: "no-hide-descendants",
          pointerEvents: "none",
        })}
        style={[style, styles?.root]}
        onAccessibilityEscape={onClose}
        onKeyDown={handleKeyDown}
        {...props}
      >
        {!hideScrim && (
          <TouchableWithoutFeedback
            accessibilityElementsHidden
            importantForAccessibility="no-hide-descendants"
            onPress={onClose}
          >
            <ModalScrim
              accessibilityHidden
              focusable={false}
              style={{ opacity }}
            />
          </TouchableWithoutFeedback>
        )}
        <FocusTrap
          disableAutoFocus={disableAutoFocus}
          disableEnforceFocus={disableEnforceFocus}
          disableRestoreFocus={disableRestoreFocus}
          enabled={isTopModal}
          open={open}
        >
          {React.cloneElement(children, {
            focusable: children.props.focusable ?? false,
            tabIndex: children.props.tabIndex ?? -1,
          })}
        </FocusTrap>
      </ModalRoot>
    </Portal>
  )
}) as OverridableComponent<ModalTypeMap>

if (__DEV__) {
  Modal.displayName = "Modal"
}
