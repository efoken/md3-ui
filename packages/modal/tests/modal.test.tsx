import { render, screen } from "@md3-ui/test-utils"
import * as React from "react"
import { Text, View } from "react-native"
import { Modal } from "../src/modal"

describe("<Modal />", () => {
  let container: HTMLDivElement

  beforeEach(() => {
    container = document.createElement("div")
    document.body.append(container)
  })

  afterEach(() => {
    container.remove()
  })

  it("should render", () => {
    render(
      <Modal open>
        <View />
      </Modal>,
      { container },
    )

    expect(document.body).toMatchSnapshot()
  })

  describe("prop: open", () => {
    it("should not render the children by default", () => {
      render(
        <Modal open={false}>
          <Text testID="content">Hello World</Text>
        </Modal>,
      )

      expect(screen.queryByTestId("content")).toBeNull()
    })

    it("should render the children inside a div through a portal when open", () => {
      render(
        <Modal open testID="portal">
          <Text>Hello World</Text>
        </Modal>,
      )

      expect(screen.getByTestId("portal")).toHaveProperty("tagName", "DIV")
    })

    it("should make the child focusable without adding a role", () => {
      render(
        <Modal open>
          <Text testID="child">Hello World</Text>
        </Modal>,
      )

      expect(screen.getByTestId("child")).not.toHaveAttribute("role")
      expect(screen.getByTestId("child")).toHaveAttribute("tabindex", "-1")
    })
  })

  describe("prop: disablePortal", () => {
    it("should render the content into the parent", () => {
      render(
        <View testID="parent">
          <Modal open disablePortal>
            <View testID="child" />
          </Modal>
        </View>,
      )
      expect(
        screen.getByTestId("parent").querySelector('[data-testid="child"]'),
      ).not.toBeNull()
    })
  })

  describe("prop: keepMounted", () => {
    it("should keep the children in the DOM", () => {
      render(
        <Modal open={false} keepMounted>
          <Text testID="content">Hello World</Text>
        </Modal>,
        { container },
      )

      expect(screen.getByTestId("content")).not.toBeNull()
    })

    it("should not include the children in the a11y tree", () => {
      const modalRef = React.createRef<View>()
      const { rerender } = render(
        <Modal ref={modalRef} keepMounted open={false}>
          <Text>ModalContent</Text>
        </Modal>,
        { container },
      )
      expect(modalRef.current).toHaveAttribute("aria-hidden", "true")

      rerender(
        <Modal ref={modalRef} keepMounted open>
          <Text>ModalContent</Text>
        </Modal>,
      )
      expect(modalRef.current).not.toHaveAttribute("aria-hidden", "true")
    })
  })
})
