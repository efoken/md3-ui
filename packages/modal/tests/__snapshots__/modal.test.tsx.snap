// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render 1`] = `
<body
  style="overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="css-view-175oi2r r-bottom-1p0dtai r-left-1d2f490 r-position-1xcajam r-right-zchlnj r-top-ipm5af r-zIndex-1prsdax"
    data-class="Modal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73 r-backgroundColor-f9p2k4 r-height-1pi2tsx r-position-u8s1d r-width-13qz1uu r-zIndex-1wyyakw"
      data-class="Modal-scrim"
      style="opacity: 1;"
      tabindex="-1"
    />
    <div
      data-testid="sentinel-start"
      tabindex="0"
    />
    <div
      class="css-view-175oi2r"
      tabindex="-1"
    />
    <div
      data-testid="sentinel-end"
      tabindex="0"
    />
  </div>
</body>
`;
